mithril-aggregator/src/services/certifier/interface.rs

```diff
@@ -0,0 +1,158 @@
+    /// the open message does not exist or the open message has been certified
+    /// since then, an error is returned.
+    async fn register_single_signature(
+        &self,
+        signed_entity_type: &SignedEntityType,
+        signature: &SingleSignatures,
+    ) -> StdResult<SignatureRegistrationStatus>;
+
+    /// Create an open message at the given beacon. If the open message does not
+    /// exist or exists at an older beacon, the older open messages are cleared
+    /// along with their associated single signatures and the new open message
+    /// is created. If the message already exists, an error is returned.
+    async fn create_open_message(
+        &self,
+        signed_entity_type: &SignedEntityType,
+        protocol_message: &ProtocolMessage,
+    ) -> StdResult<OpenMessage>;
+
+    /// Return the open message at the given Beacon. If the message does not
+    /// exist, None is returned.
+    async fn get_open_message(
+        &self,
+        signed_entity_type: &SignedEntityType,
+    ) -> StdResult<Option<OpenMessage>>;
+
+    /// Mark the open message if it has expired.
+    async fn mark_open_message_if_expired(
+        &self,
+        signed_entity_type: &SignedEntityType,
+    ) -> StdResult<Option<OpenMessage>>;
+
+    /// Create a certificate if possible. If the pointed open message does
+    /// not exist or has been already certified, an error is raised. If a multi
+    /// signature is created then the flag `is_certified` of the open
+    /// message is set to true. The Certificate is created.
+    /// If the stake quorum of the single signatures is
+    /// not reached for the multisignature to be created, the certificate is not
+    /// created and None is returned. If the certificate can be created, the
+    /// list of the registered signers for the given epoch is used.
+    async fn create_certificate(
+        &self,
+        signed_entity_type: &SignedEntityType,
+    ) -> StdResult<Option<Certificate>>;
+
+    /// Returns a certificate from its hash.
+    async fn get_certificate_by_hash(&self, hash: &str) -> StdResult<Option<Certificate>>;
+
+    /// Returns the list of the latest created certificates.
+    async fn get_latest_certificates(&self, last_n: usize) -> StdResult<Vec<Certificate>>;
+
+    /// Verify the certificate chain and epoch gap. This will return an error if
+    /// there is at least an epoch between the given epoch and the most recent
+    /// certificate.
+    async fn verify_certificate_chain(&self, epoch: Epoch) -> StdResult<()>;
+}
+
+/// ## BufferedSingleSignatureStore
+///
+/// Allow to buffer single signatures for later use when an open message isn't available yet.
+#[cfg_attr(test, mockall::automock)]
+#[async_trait]
+pub trait BufferedSingleSignatureStore: Sync + Send {
+    /// Buffer a single signature for later use.
+    async fn buffer_signature(
+        &self,
+        signed_entity_type_discriminants: SignedEntityTypeDiscriminants,
```

Comment:
Not essential:
```suggestion
        discriminant: SignedEntityTypeDiscriminants,
```
or

```suggestion
        signed_entity_type_discriminant: SignedEntityTypeDiscriminants,
```

If you think it's relevant, apply it to the following functions.